Swiss tables design for `Dict`

petvana opened the pull request:
This PR introduces simplified Swiss tables design for `Dict` described at 
~~This extends my previous PR #44332 by using Swiss tables design described at~~
https://abseil.io/about/design/swisstables#swiss-tables-design-notes

The performance gain starts to be really interesting, especially for abstract types.  I created a separate PR because the Swiss Tables can be implemented independently on #44332 but probably with less performance gian. Changes related only to the Swiss tables are in the commit [Swiss tables based hashing](https://github.com/JuliaLang/julia/commit/ae9483c5cde4356af23926ad920dfb314d0f0da9).

The main idea is to store the 7 highest bits of the hash in the slots. They can be utilized to test if the key may be equal. This limits `isequal` calls almost to a single one per operation (using a high-quality hashing function).

CPU: Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz
All times are in seconds.

**Master** (total time **52.320 s**)
<div class="data-frame"><table class="data-frame"><thead><tr><th></th><th>Type</th><th>SET</th><th>GET</th><th>GET!empty</th><th>GET!full</th><th>ITERATE</th></tr></thead><tbody><tr><th>1</th><td>Dict{Int64, Int64}</td><td>0.702</td><td>0.445</td><td>0.718</td><td>0.507</td><td>0.048</td></tr><tr><th>2</th><td>Dict{Any, Int64}</td><td>5.155</td><td>1.242</td><td>4.303</td><td>1.52</td><td>0.048</td></tr><tr><th>3</th><td>Dict{Int64, Any}</td><td>1.412</td><td>1.345</td><td>2.546</td><td>1.765</td><td>1.342</td></tr><tr><th>4</th><td>Dict{Any, Any}</td><td>5.801</td><td>3.472</td><td>6.117</td><td>3.755</td><td>0.934</td></tr><tr><th>5</th><td>Dict{String, Int64}</td><td>3.321</td><td>1.373</td><td>2.915</td><td>1.485</td><td>0.049</td></tr></tbody></table></div>

**PR**  (total time **35.070 s**)
<div class="data-frame"><table class="data-frame"><thead><tr><th></th><th>Type</th><th>SET</th><th>GET</th><th>GET!empty</th><th>GET!full</th><th>ITERATE</th></tr></thead><tbody><tr><th>1</th><td>Dict{Int64, Int64}</td><td>0.637</td><td>0.381</td><td>0.638</td><td>0.486</td><td>0.049</td></tr><tr><th>2</th><td>Dict{Any, Int64}</td><td>2.33</td><td>0.532</td><td>2.85</td><td>0.684</td><td>0.057</td></tr><tr><th>3</th><td>Dict{Int64, Any}</td><td>1.571</td><td>1.156</td><td>2.388</td><td>1.258</td><td>0.885</td></tr><tr><th>4</th><td>Dict{Any, Any}</td><td>4.239</td><td>1.213</td><td>5.671</td><td>1.245</td><td>0.888</td></tr><tr><th>5</th><td>Dict{String, Int64}</td><td>1.991</td><td>0.805</td><td>1.953</td><td>1.104</td><td>0.056</td></tr></tbody></table></div>

**PR together with #44332**  (total time **33.552 s**)
<div class="data-frame"><table class="data-frame"><thead><tr><th></th><th>Type</th><th>SET</th><th>GET</th><th>GET!empty</th><th>GET!full</th><th>ITERATE</th></tr></thead><tbody><tr><th>1</th><td>Dict{Int64, Int64}</td><td>0.597</td><td>0.356</td><td>0.59</td><td>0.444</td><td>0.051</td></tr><tr><th>2</th><td>Dict{Any, Int64}</td><td>2.788</td><td>0.506</td><td>2.278</td><td>0.592</td><td>0.053</td></tr><tr><th>3</th><td>Dict{Int64, Any}</td><td>0.773</td><td>1.113</td><td>2.376</td><td>1.091</td><td>1.66</td></tr><tr><th>4</th><td>Dict{Any, Any}</td><td>3.461</td><td>1.127</td><td>4.876</td><td>1.119</td><td>2.119</td></tr><tr><th>5</th><td>Dict{String, Int64}</td><td>1.815</td><td>0.711</td><td>1.908</td><td>1.084</td><td>0.063</td></tr></tbody></table></div>

**SwissDict** (total time **43.458 s**) from [DataStructures.jl](https://github.com/JuliaCollections/DataStructures.jl)
<div class="data-frame"><table class="data-frame"><thead><tr><th></th><th>Type</th><th>SET</th><th>GET</th><th>GET!empty</th><th>GET!full</th><th>ITERATE</th></tr></thead><tbody><tr><th>1</th><td>SwissDict{Int64, Int64}</td><td>0.776</td><td>0.359</td><td>0.813</td><td>0.459</td><td>0.018</td></tr><tr><th>2</th><td>SwissDict{Any, Int64}</td><td>3.955</td><td>1.946</td><td>3.867</td><td>1.355</td><td>0.018</td></tr><tr><th>3</th><td>SwissDict{Int64, Any}</td><td>1.257</td><td>1.282</td><td>2.105</td><td>1.899</td><td>1.328</td></tr><tr><th>4</th><td>SwissDict{Any, Any}</td><td>5.717</td><td>1.636</td><td>6.731</td><td>1.66</td><td>0.913</td></tr><tr><th>5</th><td>SwissDict{String, Int64}</td><td>1.715</td><td>1.061</td><td>1.513</td><td>1.055</td><td>0.018</td></tr></tbody></table></div>
 
<details>
<summary><b><u>Testing code</u></b></summary>

``` julia
module TestDict

using Printf
using Random
using DataStructures
using DataFrames

const n = 10_000_000

function test_set(dict, x)
    xn = length(x)
    #sizehint!(dict, xn)
    for i in 1:xn
        dict[x[i]] = i
    end
end

test_get(dict, x) = sum(dict[x[i]] for i = 1:length(x))
test_get!(dict, x) = sum(get!(dict, x[i], i) for i = 1:length(x))
test_iterate(dict) = sum(v for v = values(dict))

for D in [SwissDict, Dict]
    df = DataFrame()
    for (A,B) in [(Int, Int), (Any, Int), (Int, Any), (Any, Any), (String, Int)]
        Random.seed!(42)
        if A == String
            keys = [randstring() for i = 1:n]
        else
            keys = rand(A == Any ? Int : A, n)
        end
        keys = unique(keys)
        correct_sum = sum([1:length(keys)...])

        dict = D{A, B}()
        test_set(dict, keys)
        test_get(dict, keys)
        dict = D{A, B}()
        time_set = @elapsed test_set(dict, keys)
        time_get = @elapsed getsum = test_get(dict, keys)
        @assert getsum == correct_sum

        dict = D{A, B}()
        test_get!(dict, keys)
        test_get(dict, keys)

        dict = D{A, B}()
        time_get!_empty = @elapsed getsum = test_get!(dict, keys)
        @assert getsum == correct_sum
        time_get!_full = @elapsed getsum = test_get!(dict, keys)
        @assert getsum == correct_sum

        test_iterate(dict)
        time_iterate = @elapsed getsum = test_iterate(dict)
        @assert getsum == correct_sum

        new_data = ( 
            Type = typeof(dict), 
            SET = time_set, 
            GET = time_get, 
            GET!empty = time_get!_empty,
            GET!full = time_get!_full,
            ITERATE = time_iterate,
        )
        println(new_data)
        push!(df, new_data)
    end
    total = sum(sum(x) for x in eachcol(df[:,2:end]))
    df[:,2:end] = round.(df[:,2:end]; digits = 3)
    show(stdout, MIME("text/plain"), df)
    println("\n")
    show(stdout, MIME("text/html"), df; eltypes = false, summary = false)
    println("\n")
    @printf "Total time %.3f s\n\n\n" total
end

end
```
</details>

petvana committed:
Improve performance of 'Dict'

petvana committed:
Fix Set

petvana committed:
Fix most of the tests

petvana committed:
Use explicite types for pairs

petvana committed:
Mark broken test of precompilation

petvana committed:
Merge branch 'master' into pv-dict

petvana committed:
Reenable broken test

petvana committed:
Use simplified OldDict for testing precompilation

petvana committed:
Fix precompile test

petvana committed:
Fix white space

petvana committed:
Merge branch 'master' into pv-dict

petvana committed:
Apply suggestions from code review

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

petvana committed:
Apply suggestions from code review

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

petvana committed:
Fix performance of Set constructor if both {T,V} are bitstypes

petvana committed:
Apply suggestions from code review

petvana committed:
Add inbounds

petvana committed:
Swiss tables based hashing

petvana committed:
Fix whitespaces

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

petvana committed:
Fix Set for Swiss tables

petvana commented:
@KristofferC I'm sorry I forget to update `set.jl`. So, this nanosoldier run can be terminated.

vtjnash commented:
Would be great to hear how this compares in implementation and performance to https://juliacollections.github.io/DataStructures.jl/latest/swiss_dict/
https://nextjournal.com/eulerkochy/gsoc-20-in-datastructures.jl
@eulerkochy

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/44513#issuecomment-1061660342):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(setenv(`git push`; dir="/nanosoldier/workdir/NanosoldierReports"), ProcessExited(1)) [1]

```
Unfortunately, the logs could not be uploaded.

vtjnash commented:
https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/27b53cf_vs_dc45d77/report.md

oscardssmith commented:
Does this PR generate similar code to the LLVM-calls from the JuliaCollections dict? If so, that's really cool!

petvana commented:
> Would be great to hear how this compares in implementation and performance to https://juliacollections.github.io/DataStructures.jl/latest/swiss_dict/ https://nextjournal.com/eulerkochy/gsoc-20-in-datastructures.jl @eulerkochy

Thank you for the comment. I've added `SwissDict` into the comparison. I'm quite surprised that the PR seems to be faster (at least for abstract types). However, these two implementations are very different because the PR utilizes only one idea to store part of the hash in Metadata (here `slots`).

petvana commented:
> Does this PR generate similar code to the LLVM-calls from the JuliaCollections dict? If so, that's really cool!

@oscardssmith  Unfortunately the PR is not that cool. :-) The idea was to keep it as simple as possible and in pure Julia. It just take advantage that iterating over `Vector{UInt8}` is fast. Therefore, if you store part of the hash (7 bits) in `slots` you are able with `127/128` probability check if the key is equal. As a result, the linear probing is fast and number of `isequal` calls is limited almost to a single call per operation (~1.008 calls for optimal hashing function).

petvana committed:
Update comments + rename _shorthash7 function

oscardssmith commented on base/dict.jl:
```diff
@@ -325,23 +326,28 @@ function ht_keyindex2!(h::Dict{K,V}, key) where V where K
             end
-        elseif key === keys[index] || isequal(key, keys[index])
-            return index
+        else
```
`elseif`?

petvana commented on base/dict.jl:
```diff
@@ -352,17 +358,16 @@ function ht_keyindex2!(h::Dict{K,V}, key) where V where K
-    sz = length(h.keys)
+    sz = length(h.pairs)
     # Rehash now if necessary
     if h.ndel >= ((3*sz)>>2) || h.count*3 > sz*2
```
Thank you for the question. Generally yes, but here the primary motivation is different. The PR limits the number of `isequal` calls, and thus limits the number of allocations for abstract types and pressure on GC. These coeficients should be updated to limit memory consumption. This fine-tuning needs much more benchmarking on various sizes. There will always be some tradeoff between speed and used memory. I propose to move such a discussion into a separate PR. Meanwhile, I'll try to prepare some microbenchmarks.

petvana replied:
What I've tested so far, we would need to use SIMD (as in [DataStructures.jl](https://github.com/JuliaCollections/DataStructures.jl)) to increase the capacity without a significant performance drop (given by unpredictable branching). The good news is, that metadata in `slots` is prepared for that. Nevertheless, I'm not sure, if such a low-level llvm code should go to base because it will be harder to read and check ... and it will be platform specific.

petvana committed:
Apply suggestions from code review

oscardssmith commented:
I think this is basically ready to merge. Can you add a benchmark for iterating over a `Dict` to the suite? That looks like the main type of test missing.

oscardssmith commented:
One other optimization that we should either include here or in a followup PR is that for `get` we should do a linear probe over `slots`for dictionaries with 32 or fewer elements. The linear lookup should be easy for LLVM to vectorize, and should be simpler and faster.

JeffBezanson commented on base/dict.jl:
```diff
@@ -87,14 +87,13 @@ mutable struct Dict{K,V} <: AbstractDict{K,V}
     function Dict{K,V}() where V where K
         n = 16
-        new(zeros(UInt8,n), Vector{K}(undef, n), Vector{V}(undef, n), 0, 0, 0, 1, 0)
+        new(fill(0x80,n), Vector{Pair{K,V}}(undef, n), 0, 0, 0, 1, 0)
```
Would it make sense to have 0x00 represent empty, so we can start with zero'd memory in case that ever becomes especially fast or convenient?

petvana replied:
Probably good idea, but it should have no effect for now. I'll try that. The coding of metadata would be:
``` julia
# Metadata: empty => 0x00, removed => 0x7f, full => 0b1[7 most significant hash bits]
```

petvana replied:
Updated and there is no significant performance difference, only a single extra `|` operation.
https://github.com/JuliaLang/julia/blob/8fd9617f7906ec78cde6d867e43857a64aa56eff/base/dict.jl#L170-L172

JeffBezanson commented on base/dict.jl:
```diff
@@ -279,44 +281,43 @@ end
-        iter > maxprobe && break
+        (iter += 1) > maxprobe && return -1
     end
     return -1
```
This can be replaced with a comment that this is unreachable.

fredrikekre commented:
I guess the breakage of this PR is the same as #44332, but lets check:

@nanosoldier `runtests(ALL, vs = ":master")`

petvana committed:
Remove unreachable code

JeffBezanson commented:
I like the idea of storing a vector of pairs, but it occurs to me this can have a large cost in alignment padding, for example in a `Dict{Int64, Int8}`.

petvana commented:
> I like the idea of storing a vector of pairs, but it occurs to me this can have a large cost in alignment padding, for example in a `Dict{Int64, Int8}`.

This is a design choice and I'm definitely the right one in this conversation to decide. I'll benchmark such combinations. I see the following options:

1.  Merge as it is (both vector of pairs and Swiss table design) - Slightly braking (for example single change in CSV.jl)
2.  Split the PR and merge only the Swiss table design. - Almost no breakage
3.  Split the PR, merge only the Swiss table design now, and merge a vector of pairs to Julia 2.0. -  The most breaking change would be postponed to a major release.

oscardssmith commented:
Assuming benchmarks for option 2 look good, that's probably what I would want. We definitely wouldn't wait to do `Vector{Pair}` for Julia 2.0. It's not breaking, so if it's better we'll merge it nowish, and if it's worse, we won't merge it for 2.0. I also discovered that Google has an Apache licensed implementation of swish hash [here](https://github.com/abseil/abseil-cpp/tree/master/absl/hash).

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/44513#issuecomment-1065575530) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/99e24d7_vs_5ec1f9f/report.html).

KristofferC commented:
Personally, I like keeping fairly orthogonal things in different PRs so to me, only focusing on the Swiss here and having it be merged relatively quickly is advantageous over bundling it together with the `Pair` discussion that has to also take into account the padding and "breakage".  So, in my opinion, make this PR only have the Swiss and rebase the other one on top of this.

petvana committed:
Focus PR purely on Swiss table design

petvana committed:
Make empty slot 0x00

petvana committed:
Improve backward compatibility

petvana committed:
Merge branch 'master' into pv-dict-swiss

KristofferC commented:
Great and thanks for putting in the extra effort of keeping things backward compatible! 

IMO this is mergeable but maybe @JeffBezanson wants to look it over one last time.

petvana committed:
Minor change (revert fill() to zeros())

petvana commented:
I've gone throw the code for the last time and reverted a single line (`fill` to `zeros`). Thus, ready to be merged from my point of view.
